### ISSUE TITLE
Implement `groupBy` operator

### ISSUE THREAD

**Issue by mihaisoloi** [monifu/monifu#56]
Another important operator is the `groupBy` operator, this allows for the split of an Observable[Map[String, (String, String)] into multiple Observable[(String, String)]. You can find a more detailed explanation here: http://reactivex.io/documentation/operators/groupby.html

**Comment by alexandru** [comment]
Sorry, I took a look and we're not going to implement this in Monifu. This is because we can't do back-pressure with `groupBy`.

The problem is that this operator ends up taking data from the source observable and pushing that data into multiple observers, which are either the corresponding input channels of our `GroupedObservables` and that are created dynamically per key, or the downstream observer that is going to get these `GroupedObservables`. So we've got 2 issues:

1. we cannot apply back-pressure on the downstream observer's response, because that would also block the streaming of events into our grouped streams
2. the inputs have to be some sort of buffers that don't lose events until subscribed - but unfortunately we can't stream `ConnectableObservable`, because that's error-prone - so we must come up with an Observable type that's neither cold nor hot, an Observable that caches events until subscribed and that can be subscribed only once - this is what RxJava does and frankly, it smells
3. I have yet to find a use-case for `groupBy` that can't be solved by other more sane operators
